### PR TITLE
⏩ Extend renderers to allow for kind/subtypes

### PR DESCRIPTION
**Pull request by rowanc1**
This allows for multiple renderers to be attached to a single node, based on a match of the kind.

For example you can now add a specific renderer for a `link[kind=github]`. This allows for inheriting these renderers and then extending or overriding them independently.

### CODE REVIEW

File: packages/myst-to-react/src/links/index.tsx [hunk 1]
```diff
@@ -134,8 +137,20 @@ export const linkBlock: NodeRenderer<TransformedLink> = ({ node }) => {
...
+const LINK_RENDERERS: NodeRenderers = {
+  link: {
+    base: SimpleLink,
+    // Then duplicate the renderers for protocols
+    'link[protocol=github]': GithubLinkRenderer,
+    'link[protocol=wiki]': WikiLinkRenderer,
+    'link[protocol=rrid]': RRIDLinkRenderer,
+    'link[protocol=ror]': RORLinkRenderer,
+    // Put the kinds last as they will match first in the future
+    'link[kind=github]': GithubLinkRenderer,
```
Comment (rowanc1) [inline packages/myst-to-react/src/links/index.tsx:149]: Prepares for:

https://github.com/jupyter-book/mystmd/issues/1417

File: packages/myst-to-react/src/MyST.tsx [hunk 2]
```diff
@@ -10,17 +12,25 @@ function DefaultComponent({ node }: { node: GenericNode }) {
   );
 }
 
+export function selectRenderer(renderers: NodeRenderersValidated, node: GenericNode) {
+  const componentRenderers = renderers[node.type] ?? renderers['DefaultComponent'];
+  const SpecificComponent = Object.entries(componentRenderers)
+    .reverse()
```
Comment (fwkoch) [inline packages/myst-to-react/src/MyST.tsx:18]: Why `reverse`?

File: packages/providers/src/renderers.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      throw new Error(
+        `Renderer for "${key}" must be either a function or an object containing a "base" renderer.`,
+      );
+    }
+  }
+
+  return validatedRenderers;
+}
+
+export function mergeRenderers(renderers: NodeRenderers[], validate: true): NodeRenderersValidated;
```
Comment (fwkoch) [inline packages/providers/src/renderers.tsx:34]: A few more docstrings - particularly for the functions in this file - would be helpful.

File: packages/myst-to-react/src/MyST.tsx [hunk 4]
```diff
@@ -10,17 +12,25 @@ function DefaultComponent({ node }: { node: GenericNode }) {
   );
 }
 
+export function selectRenderer(renderers: NodeRenderersValidated, node: GenericNode) {
+  const componentRenderers = renderers[node.type] ?? renderers['DefaultComponent'];
+  const SpecificComponent = Object.entries(componentRenderers)
```
Comment (fwkoch) [inline packages/myst-to-react/src/MyST.tsx:17]:
```suggestion
  const specificComponent = Object.entries(componentRenderers)
```

File: packages/myst-to-react/src/links/index.tsx [hunk 5]
```diff
@@ -134,8 +137,20 @@ export const linkBlock: NodeRenderer<TransformedLink> = ({ node }) => {
   );
 };
 
-const LINK_RENDERERS = {
-  link,
+const LINK_RENDERERS: NodeRenderers = {
+  link: {
```
Comment (fwkoch) [inline packages/myst-to-react/src/links/index.tsx:141]: I don't quite see why we need this nested structure as opposed to:
```ts
{
  link: SimpleLink,
  'link[protocol=github]': GithubLinkRenderer,
  'link[protocol=wiki]': WikiLinkRenderer,
  ...
}
```

I guess the nesting lets us first narrow down based on `type` so we are not doing `selector` checks against _all_ the `renderers`?

Comment (fwkoch) [reply packages/myst-to-react/src/links/index.tsx:141]: My next question is: would this work the same if we did something like:
```ts
{
  link: {
    base: SimpleLink,
    '[protocol=github]': GithubLinkRenderer,
    '[protocol=wiki]': WikiLinkRenderer,
    ...
  }
}
```

i.e. we don't need to repeat `"link"` since we are only dealing with `link` nodes at this point?

**Review by fwkoch — COMMENTED** [review]
I think this makes sense. Now, instead of a simple `type: renderer` lookup, we have a more complex `type: { renderers }` structure where the basic renderer is under `base` and other renderers are under `selector` keys.

Keying off `selector` strings seems a little complicated (as opposed to, say, just `kind` or something) but it is quite flexible and powerful. Some of this logic is a little hard to parse without a bit more docstring-level documentation, though.
